File: libs/store-dao/src/main/kotlin/mono/store/dao/workspace/WorkspaceObjectDao.kt
/*
 * Copyright (c) 2023, tuanchauict
 */

package mono.store.dao.workspace

import mono.common.currentTimeMillis
import mono.graphics.geo.Point
import mono.shape.serialization.SerializableGroup
import mono.shape.serialization.ShapeSerializationUtil
import mono.store.manager.StorageDocument
import mono.store.manager.StoreKeys.OBJECT_CONTENT
import mono.store.manager.StoreKeys.OBJECT_LAST_MODIFIED
import mono.store.manager.StoreKeys.OBJECT_LAST_OPENED
import mono.store.manager.StoreKeys.OBJECT_NAME
import mono.store.manager.StoreKeys.OBJECT_OFFSET

/**
 * A dao for an object (aka project or file) in the workspace.
 */
class WorkspaceObjectDao internal constructor(
    val objectId: String,
    workspaceDocument: StorageDocument
) {

    private val objectDocument: StorageDocument = workspaceDocument.childDocument(objectId)

    var offset: Point
        get() {
            val offsetString = objectDocument.get(OBJECT_OFFSET)
            val (leftString, topString) = offsetString?.split('|')?.takeIf { it.size == 2 }
                ?: return Point.ZERO
            val left = leftString.toIntOrNull() ?: return Point.ZERO
            val top = topString.toIntOrNull() ?: return Point.ZERO
            return Point(left, top)
        }
        set(value) = objectDocument.set(OBJECT_OFFSET, "${value.left}|${value.top}")

    var rootGroup: SerializableGroup?
        get() {
            val json = objectDocument.get(OBJECT_CONTENT) ?: return null
            return ShapeSerializationUtil.fromJson(json) as? SerializableGroup
        }
        set(value) {
            if (value != null) {
                val json = ShapeSerializationUtil.toJson(value)
                objectDocument.set(OBJECT_CONTENT, json)
            }
            lastModifiedTimestampMillis = currentTimeMillis()
        }

    var name: String
        get() = objectDocument.get(OBJECT_NAME) ?: DEFAULT_NAME
        set(value) {
            objectDocument.set(OBJECT_NAME, value)
            lastModifiedTimestampMillis = currentTimeMillis()
        }

    var lastModifiedTimestampMillis: Double
        get() {
            val lastModifiedString = objectDocument.get(OBJECT_LAST_MODIFIED)
            return lastModifiedString?.toDoubleOrNull() ?: currentTimeMillis()
        }
        private set(value) = objectDocument.set(OBJECT_LAST_MODIFIED, value.toString())

    var lastOpened: Double
        get() {
            val lastOpenedString = objectDocument.get(OBJECT_LAST_OPENED)
            return lastOpenedString?.toDoubleOrNull() ?: currentTimeMillis()
        }
        private set(value) = objectDocument.set(OBJECT_LAST_OPENED, value.toString())

    fun updateLastOpened() {
        lastOpened = currentTimeMillis()
    }

    fun removeSelf() {
        with(objectDocument) {
            remove(OBJECT_OFFSET)
            remove(OBJECT_CONTENT)
            remove(OBJECT_NAME)
            remove(OBJECT_LAST_MODIFIED)
        }
    }

    companion object {
        const val DEFAULT_NAME = "Undefined"
    }
}

File: monosketch-svelte/src/lib/mono/store-manager/dao/workspace-object-dao.ts
import { Point } from "$libs/graphics-geo/point";
import { SerializableLineConnector } from "$mono/shape/serialization/connector";
import { SerializableGroup } from "$mono/shape/serialization/shapes";
import { StorageDocument, StoreKeys } from "$mono/store-manager";

/**
 * A dao for an object (aka project or file) in the workspace.
 */
export class WorkspaceObjectDao {
    private objectDocument: StorageDocument;

    constructor(public objectId: string, private workspaceDocument: StorageDocument) {
        this.objectDocument = workspaceDocument.childDocument(objectId);
    }

    get offset(): Point {
        const storedValue = this.objectDocument.get(StoreKeys.OBJECT_OFFSET);
        if (storedValue === null) {
            return Point.ZERO;
        }
        const [left, top] = storedValue.split(',').map(parseFloat);
        if (isNaN(left) || isNaN(top)) {
            return Point.ZERO;
        }
        return Point.of(left, top);
    }

    set offset(value: Point) {
        this.objectDocument.set(StoreKeys.OBJECT_OFFSET, `${value.left},${value.top}`);
    }

    get rootGroup(): SerializableGroup | null {
        const storedValue = this.objectDocument.get(StoreKeys.OBJECT_CONTENT);
        if (storedValue === null) {
            return null;
        }
        // @ts-expect-error fromJson
        return SerializableGroup.fromJson(JSON.parse(storedValue));
    }

    set rootGroup(value: SerializableGroup | null) {
        if (value !== null) {
            // @ts-expect-error toJson
            const json = value.toJson();
            this.objectDocument.set(StoreKeys.OBJECT_CONTENT, JSON.stringify(json));
        }
        this.lastModifiedTimestampMillis = Date.now();
    }

    get connectors(): SerializableLineConnector[] {
        const storedValue = this.objectDocument.get(StoreKeys.OBJECT_CONNECTORS);
        if (storedValue === null) {
            return [];
        }
        const jsonArray = JSON.parse(storedValue);
        // @ts-expect-error fromJson
        return jsonArray.map((json) => SerializableLineConnector.fromJson(json));
    }

    set connectors(value: SerializableLineConnector[]) {
        // @ts-expect-error toJson
        const jsonArray = value.map(connector => connector.toJson());
        this.objectDocument.set(StoreKeys.OBJECT_CONNECTORS, JSON.stringify(jsonArray));
        this.lastModifiedTimestampMillis = Date.now();
    }

    get name(): string {
        return this.objectDocument.get(StoreKeys.OBJECT_NAME) ?? DEFAULT_NAME;
    }

    set name(value: string) {
        this.objectDocument.set(StoreKeys.OBJECT_NAME, value);
        this.lastModifiedTimestampMillis = Date.now();
    }

    get lastModifiedTimestampMillis(): number {
        const storedValue = this.objectDocument.get(StoreKeys.OBJECT_LAST_MODIFIED);
        if (storedValue === null) {
            return Date.now();
        }
        return parseFloat(storedValue);
    }

    set lastModifiedTimestampMillis(value: number) {
        this.objectDocument.set(StoreKeys.OBJECT_LAST_MODIFIED, value.toString());
        this.workspaceDocument.set(StoreKeys.LAST_MODIFIED_TIME, value.toString());
        this.workspaceDocument.set(StoreKeys.LAST_MODIFIED_PROJECT_ID, this.objectId);
    }

    get lastOpened(): number {
        const storedValue = this.objectDocument.get(StoreKeys.OBJECT_LAST_OPENED);
        if (storedValue === null) {
            return Date.now();
        }
        return parseFloat(storedValue);
    }

    updateLastOpened() {
        this.objectDocument.set(StoreKeys.OBJECT_LAST_OPENED, Date.now().toString());
        this.workspaceDocument.set(StoreKeys.LAST_MODIFIED_PROJECT_ID, this.objectId);
    }

    removeSelf(): void {
        const storeKeys = [
            StoreKeys.OBJECT_OFFSET,
            StoreKeys.OBJECT_CONTENT,
            StoreKeys.OBJECT_CONNECTORS,
            StoreKeys.OBJECT_NAME,
            StoreKeys.OBJECT_LAST_MODIFIED,
            StoreKeys.OBJECT_LAST_OPENED,
        ];
        for (const key of storeKeys) {
            this.objectDocument.remove(key);
        }
    }
}

export const DEFAULT_NAME = "Undefined";
